spateo/tools/cluster/leiden.py

```diff
@@ -81,23 +81,43 @@ def calculate_leiden_partition(
     Returns:
         clusters: Array containing cluster assignments
     """
+    from ...logging import logger_manager as lm
+
+    logger = lm.get_main_logger()
     if adj is None and input_mat is None:
         raise ValueError("Either `adj` or `input_mat` must be specified")
 
+    logger.info("using adj_matrix from arg for clustering...")
+
     if adj is not None:
-        if isinstance(adj, np.ndarray):
-            G = adj_to_igraph(adj.tolist())
+        if scipy.sparse.issparse(adj):
+            pass
         else:
-            G = igraph.Graph(n=adj.shape[0])
-            for i, j in zip(*adj.nonzero()):
-                G.add_edge(i, j)
+            adj = scipy.sparse.csr_matrix(adj)
+        sources, targets = adj.nonzero()
+        weights = adj[sources, targets]
+        if isinstance(weights, np.matrix):
+            weights = weights.A1
+        G = igraph.Graph(directed=None)
+        G.add_vertices(adj.shape[0])  # this adds adjacency.shape[0] vertices
+        G.add_edges(list(zip(sources, targets)))
```

Comment:
is this function  G.add_edges(list(zip(sources, targets))) fast? is it as fast as adj_to_igraph(adj.tolist())?

Reply:
Yes, I directly utilized the properties of sparse matrices here, replacing the original for loop iteration for adding nodes, which reduced the time complexity. This not only improved the algorithm's efficiency but also reduced memory usage.

Reply:
You're right, `adj_to_igraph(adj.tolist())` would indeed be faster, but both are faster than a for loop to add nodes

Reply:
then maybe we can use igraph.Graph.Weighted_Adjacency(adj) to replace all these lines of code:

```
        if isinstance(weights, np.matrix):
            weights = weights.A1
        G = igraph.Graph(directed=None)
        G.add_vertices(adj.shape[0])  # this adds adjacency.shape[0] vertices
        G.add_edges(list(zip(sources, targets)))
```